src/macroxs_operations.F90

```diff
@@ -0,0 +1,243 @@
+! MACROXS_SAMPLE_SCATTER performs the work with ScattData to sample outgoing
+! energy and change in angle.
+!===============================================================================
+
+  subroutine macroxs_sample_scatter(scatt, gin, gout, mu, wgt)
+    class(ScattData), intent(in)    :: scatt ! Scattering Object to Use
+    integer,          intent(in)    :: gin   ! Incoming neutron group
+    integer,          intent(out)   :: gout  ! Sampled outgoin group
+    real(8),          intent(out)   :: mu    ! Sampled change in angle
+    real(8),          intent(inout) :: wgt   ! Particle weight
+
+    real(8) :: xi     ! Our random number
+    real(8) :: prob   ! Running probability
+    integer :: imu
+    real(8) :: u, f, M
+    real(8) :: mu0, frac, mu1
+    real(8) :: c_k, c_k1, p0, p1
+    integer :: k, NP, samples
+
+    xi = prn()
+    prob = ZERO
+    gout = 0
+
+    do while (prob < xi)
+      gout = gout + 1
+      prob = prob + scatt % energy(gout,gin)
+    end do
+
+    select type (scatt)
+    type is (ScattDataHistogram)
+      xi = prn()
+      if (xi < scatt % data(1,gout,gin)) then
+        imu = 1
+      else
+        imu = binary_search(scatt % data(:,gout,gin), &
+                            size(scatt % data(:,gout,gin)), xi)
+      end if
+
+      ! Randomly select a mu in this bin.
+      mu = prn() * scatt % dmu + scatt % mu(imu)
+
+    type is (ScattDataTabular)
+      ! determine outgoing cosine bin
+      NP = size(scatt % data(:,gout,gin))
+      xi = prn()
+
+      c_k = scatt % data(1,gout,gin)
+      do k = 1, NP - 1
+        c_k1 = scatt % data(k+1,gout,gin)
+        if (xi < c_k1) exit
+        c_k = c_k1
+      end do
+
+      ! check to make sure k is <= NP - 1
+      k = min(k, NP - 1)
+
+      p0  = scatt % fmu(k,gout,gin)
+      mu0 = scatt % mu(k)
+      ! Linear-linear interpolation to find mu value w/in bin.
+      p1  = scatt % fmu(k+1,gout,gin)
+      mu1 = scatt % mu(k+1)
+
+      frac = (p1 - p0)/(mu1 - mu0)
+
+      if (frac == ZERO) then
+        mu = mu0 + (xi - c_k)/p0
+      else
+        mu = mu0 + (sqrt(max(ZERO, p0*p0 + TWO*frac*(xi - c_k))) - p0)/frac
+      end if
+
+      if (mu <= -ONE) then
+        mu = -ONE
+      else if (mu >= ONE) then
+        mu = ONE
+      end if
+
+    type is (ScattDataLegendre)
+      ! Now we can sample mu using the legendre representation of the scattering
+      ! kernel in data(1:this % order)
+
+      ! Do with rejection sampling
+      ! Set upper bound (instead of searching for max - though this is inefficient)
+      M = 4.0_8
+      samples = 0
+      do
+        mu = TWO * prn() - ONE
+        f = scatt % calc_f(gin,gout,mu)
+        if (f > ZERO) then
+          u = prn() * M
+          if (u <= f) then
+            exit
+          end if
+        end if
+        samples = samples + 1
+        if (samples > MAX_SAMPLE) then
+          ! Exit with an isotropic event.
+          exit
```

Comment:
That's fine.  Can do.  Trivial now with a fatal_error routine without
circular dependency baggage :-)

On Fri, Feb 12, 2016 at 9:02 AM, Paul Romano <notifications@github.com>
wrote:

> In src/macroxs_operations.F90
> <https://github.com/mit-crpg/openmc/pull/494#discussion_r52744008>:
>
> > +      ! Set upper bound (instead of searching for max - though this is inefficient)
> > +      M = 4.0_8
> > +      samples = 0
> > +      do
> > +        mu = TWO * prn() - ONE
> > +        f = scatt % calc_f(gin,gout,mu)
> > +        if (f > ZERO) then
> > +          u = prn() * M
> > +          if (u <= f) then
> > +            exit
> > +          end if
> > +        end if
> > +        samples = samples + 1
> > +        if (samples > MAX_SAMPLE) then
> > +          ! Exit with an isotropic event.
> > +          exit
>
> This is a little disconcerting to me. Why not just throw an error rather
> than passing silently with a mu that doesn't physically make sense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/494/files#r52744008>.
>